### PR TITLE
improves secure_password in salt/utils/pycrypto.py (#59486)

### LINKED ISSUE #59486: [FEATURE REQUEST] increase character set for random.get_str()
**Is your feature request related to a problem? Please describe.**

We are generating random passwords using [`salt['random.get_str'](12)`] (https://docs.saltproject.io/en/latest/ref/modules/all/salt.modules.mod_random.html#salt.modules.mod_random.get_str) on state call for system users.
As the randomness of this method is sufficient for us the method is near to perfect to generate passwords on the fly.
The only issue is that the generated passwords only consist of alpha-numeric characters.
Some of our systems require the passwords to include symbols too, which is why we cannot use the generated string for those systems.

**Describe the solution you'd like**

In my opinion it would be an improvement for `random.get_str()` to allow controling the character space used to generate the string.

1. The character space could be given via string `salt['random.get_str'](12, chars='abcde...')` to simplify the usage a helper function could be provided that generates the list of characters based on `string`-character-set (ascii_letters, ascii_lowercase, ascii_uppercase, digits, punctuation, printable).
2. The character space could be controlled via boolean parameters like `salt['random.get_str'](12, chars_printable=True, chars_digits=True, chars_letters=True, ...)` which then activate the `string`-character-sets (ascii_letters, ascii_lowercase, ascii_uppercase, digits, punctuation, printable) accordingly.

**Describe alternatives you've considered**

We have tested the module with different supported crypto-libraries (M2Crypto, Cryptodome & Crypto) but due to [re.sub()](https://github.com/saltstack/salt/blob/11199bb5e168a2da439d2fdf8b3ab726a5392e6f/salt/utils/pycrypto.py#L59) they didn't change the behaviour.
Currently for the affected systems we are just appending hard-coded symbols to the generated string.

**Additional context**
None

**Please Note**
If this feature request would be considered a substantial change or addition, this should go through a SEP process here https://github.com/saltstack/salt-enhancement-proposals, instead of a feature request.

### PR DESCRIPTION
**Pull request by root360-AndreasUlm**
### What does this PR do?

This PR changes the default character set used by `utils.pycrypto.secure_password()` and implements arguments to control the used character set. (see #59486)

### What issues does this PR fix or reference?
Fixes: None
Closes: #59486
Updated PR for https://github.com/saltstack/salt/pull/59533

### Previous Behavior

`utils.pycrypto.secure_password()` generated random string based on `string.ascii_letters + string.digits`.

### New Behavior

`utils.pycrypto.secure_password()` generates random string based on given character controlling arguments.
The default is now `string.ascii_lowercase + string.ascii_uppercase + string.digits + string.punctuation`.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [X] Docs
- [X] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [X] Tests written/updated

### Commits signed with GPG?
Yes

### CODE REVIEW

**Comment by root360-AndreasUlm** [comment]
FYI: This PR was created with pre-commit-hooks disabled due to limitations on development system and hooks not running with containers (yet) (see https://github.com/saltstack/salt/issues/59486#issuecomment-781790748)

File: salt/modules/mod_random.py [hunk 1]
```diff
@@ -96,7 +96,7 @@ def str_encode(value, encoder="base64"):
     return out
 
 
-def get_str(length=20):
+def get_str(length=20, **kwargs):
```
Comment (s0undt3ch) [inline salt/modules/mod_random.py:99]: Please be explicit on the keyword arguments and their defaults. We don't really need this to be `**kwargs`

File: salt/modules/mod_random.py [hunk 2]
```diff
@@ -105,13 +105,63 @@ def get_str(length=20):
...
     CLI Example:
 
     .. code-block:: bash
 
         salt '*' random.get_str 128
+        salt '*' random.get_str 128 chars='abc123.!()'
+        salt '*' random.get_str 128 lowercase=False whitespace=True
     """
-    return salt.utils.pycrypto.secure_password(length)
+    return salt.utils.pycrypto.secure_password(length, **kwargs)
```
Comment (s0undt3ch) [inline salt/modules/mod_random.py:164]: as above, pass explicit keyword arguments and their values

File: salt/utils/pycrypto.py [hunk 3]
```diff
@@ -41,10 +41,26 @@
 log = logging.getLogger(__name__)
 
 
-def secure_password(length=20, use_random=True):
+def secure_password(length=20, use_random=True, **kwargs):
```
Comment (s0undt3ch) [inline salt/utils/pycrypto.py:44]: Please add the supported keyword arguments and their default values, we don't need this to be `**kwargs`

**Review by s0undt3ch — CHANGES_REQUESTED** [review]

**Comment by root360-AndreasUlm** [comment]
* force rebased the PR on master
* solved linting issues 
* solved coding comments of @s0undt3ch

**Comment by s0undt3ch** [comment]
I pushed a commit with changes to the docstring

**Comment by root360-AndreasUlm** [comment]
> I pushed a commit with changes to the docstring

nice thank you.
I haven't found a documentation about the possible keys to indicate version, yet.
Is there a documentation?

**Comment by s0undt3ch** [comment]
The version numbers? the `versionadded`, `versionchanged`?

**Comment by root360-AndreasUlm** [comment]
> The version numbers? the `versionadded`, `versionchanged`?

yes those.

**Comment by s0undt3ch** [comment]
https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-versionadded

**Review by s0undt3ch — DISMISSED** [review]

**Comment by root360-AndreasUlm** [comment]
@s0undt3ch Do all checks have to be successful and the branch always up-to-date for a valid merge request?
I'm currently facing CI-jobs being canceled and multiple times a day I get the message that the branch is out-dated.

**Comment by s0undt3ch** [comment]
> Do all checks have to be successful

Yes

> and the branch always up-to-date for a valid merge request?

Not necessarily

> I'm currently facing CI-jobs being canceled and multiple times a day I get the message that the branch is out-dated.

Yes, our Jenkins instance had problems yesterday and there's some instability in our CI jobs which we're trying to address.

**Comment by root360-AndreasUlm** [comment]
@s0undt3ch I fixed file conflicts and now all checks have passed.

**Review by s0undt3ch — DISMISSED** [review]

**Comment by s0undt3ch** [comment]
Conflicts fixed. updated `versionadded`, `versionchanged` to `3004.0`

**Review by dwoz — APPROVED** [review]
I'm good with these changes but `tests/unit/utils/test_pycrypto.py` should be removed if we're not using it.

**Pull request merged**